### PR TITLE
r/budgets_budget_action - new resource

### LINKED ISSUE #19541: AWS Budget Actions
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

We have a use case where we need to set a budget for an AWS account, member or Organizations Unit (OU), and when it hits the upper limit/threshold set, it'll notify (this feature exist) and apply SCP (policy) to that account.
This would prevent the users further launching resources, `Deny` policy action.
Seems the API is already available for [budget action](https://docs.aws.amazon.com/aws-cost-management/latest/APIReference/API_budgets_Action.html), but not yet available in Terraform AWS [provider](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/budgets_budget) `v.3.42.0`.
However, seems we can only apply the SCP to OU level. 
Would be great to apply on AWS account level as well.

### New or Affected Resource(s)

* aws_budgets_budget

### Potential Terraform Configuration

```hcl
resource "aws_budgets_budget" "ec2" {
  name              = "budget-ec2-monthly"
  budget_type       = "COST"
  limit_amount      = "1200"
  limit_unit        = "USD"
  time_period_end   = "2087-06-15_00:00"
  time_period_start = "2017-07-01_00:00"
  time_unit         = "MONTHLY"

  cost_filters = {
    Service = "Amazon Elastic Compute Cloud - Compute"
  }

  notification {
    comparison_operator        = "GREATER_THAN"
    threshold                  = 100
    threshold_type             = "PERCENTAGE"
    notification_type          = "FORECASTED"
    subscriber_email_addresses = ["test@example.com"]
  }

  budget_action {
    action_type = "APPLY_SCP_POLICY"
    approval_model = "AUTOMATIC"
    ou_ids = ["ou-dev", "ou-prod"]
  }
}
```

### References

https://docs.aws.amazon.com/aws-cost-management/latest/APIReference/API_budgets_Action.html
https://docs.aws.amazon.com/aws-cost-management/latest/APIReference/API_budgets_ScpActionDefinition.html

* #0000

### PR DESCRIPTION
**Pull request by DrFaust92**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #19541

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSBudgetsBudgetAction_'
--- PASS: TestAccAWSBudgetsBudgetAction_disappears (49.87s)
--- PASS: TestAccAWSBudgetsBudgetAction_basic (55.76s)
```

### CODE REVIEW

File: website/docs/r/budgets_budget_action.html.markdown [hunk 1]
```diff
@@ -0,0 +1,157 @@
+---
+subcategory: "Budgets"
+layout: "aws"
+page_title: "AWS: aws_budgets_budget_action"
+description: |-
+  Provides a budgets budget action resource.
```
Comment (ewbankkit) [inline website/docs/r/budgets_budget_action.html.markdown:6]:
```suggestion
  Provides a budget action resource.
```

File: website/docs/r/budgets_budget_action.html.markdown [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+---
+
+# Resource: aws_budgets_budget_action
+
+Provides a budgets budget resource. Budgets use the cost visualisation provided by Cost Explorer to show you the status of your budgets, to provide forecasts of your estimated costs, and to track your AWS usage, including your free tier usage.
```
Comment (ewbankkit) [inline website/docs/r/budgets_budget_action.html.markdown:11]:
```suggestion
Provides a budget action resource. Budget actions are cost savings controls that run either automatically on your behalf or by using a workflow approval process.
```

**Comment by ewbankkit** [comment]
```console
% make testacc TEST=./aws TESTARGS='-run=TestAccAWSBudgetsBudgetAction_'   
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSBudgetsBudgetAction_ -timeout 180m
=== RUN   TestAccAWSBudgetsBudgetAction_basic
=== PAUSE TestAccAWSBudgetsBudgetAction_basic
=== RUN   TestAccAWSBudgetsBudgetAction_disappears
=== PAUSE TestAccAWSBudgetsBudgetAction_disappears
=== CONT  TestAccAWSBudgetsBudgetAction_basic
=== CONT  TestAccAWSBudgetsBudgetAction_disappears
    resource_aws_budgets_budget_action_test.go:120: Step 1/1 error: Check failed: Check 2/2 error: Deleting Budget Action failed: ResourceLockedException: This method is not allowed during [ActionStatus: Execution_In_Progress]
--- FAIL: TestAccAWSBudgetsBudgetAction_disappears (15.55s)
=== CONT  TestAccAWSBudgetsBudgetAction_basic
    testing_new.go:63: Error running post-test destroy, there may be dangling resources: exit status 1
        2021/06/08 09:15:13 [DEBUG] Using modified User-Agent: Terraform/0.12.31 HashiCorp-terraform-exec/0.13.3
        
        Error: Deleting Budget Action failed: ResourceLockedException: This method is not allowed during [ActionStatus: Execution_In_Progress]
        
        
--- FAIL: TestAccAWSBudgetsBudgetAction_basic (19.96s)
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	23.434s
FAIL
make: *** [testacc] Error 1
```

When I then view the budget action it's in `EXECUTION_FAILURE` state and can be deleted:

```console
% aws budgets describe-budget-actions-for-account --account-id 123456789012                                    
{
    "Actions": [
        {
            "ActionId": "69be9bb5-11b0-4a67-aad0-dbd9fcb68e0d",
            "BudgetName": "tf-acc-test-1612764617238458539",
            "NotificationType": "ACTUAL",
            "ActionType": "APPLY_IAM_POLICY",
            "ActionThreshold": {
                "ActionThresholdValue": 100.0,
                "ActionThresholdType": "ABSOLUTE_VALUE"
            },
            "Definition": {
                "IamActionDefinition": {
                    "PolicyArn": "arn:aws:iam::123456789012:policy/tf-acc-test-1612764617238458539",
                    "Roles": [
                        "tf-acc-test-1612764617238458539"
                    ]
                }
            },
            "ExecutionRoleArn": "arn:aws:iam::123456789012:role/tf-acc-test-1612764617238458539",
            "ApprovalModel": "AUTOMATIC",
            "Status": "EXECUTION_FAILURE",
            "Subscribers": [
                {
                    "SubscriptionType": "EMAIL",
                    "Address": "test@test.test"
                }
            ]
        }
    ]
}
% aws budgets delete-budget-action --account-id 123456789012 --budget-name "tf-acc-test-1612764617238458539" --action-id "69be9bb5-11b0-4a67-aad0-dbd9fcb68e0d"
% aws budgets describe-budget-actions-for-account --account-id 123456789012                                                                                    
{
    "Actions": []
}
```

Some sort of `waiter` will need to be implemented.
I can't find any state transition documentation though.

**Comment by DrFaust92** [comment]
Cool, didnt experience this. will take a look regarding delete waiter.

**Comment by ewbankkit** [comment]
It may be a `Create` and/or `Update` waiter that is required.
Thanks.

**Comment by DrFaust92** [comment]
@ewbankkit added waiters but as you said there are no specific docs on this and i made a guess to what's right

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

##### Commercial

```console
% make testacc TEST=./aws TESTARGS='-run=TestAccAWSBudgetsBudgetAction_'          
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSBudgetsBudgetAction_ -timeout 180m
=== RUN   TestAccAWSBudgetsBudgetAction_basic
=== PAUSE TestAccAWSBudgetsBudgetAction_basic
=== RUN   TestAccAWSBudgetsBudgetAction_disappears
=== PAUSE TestAccAWSBudgetsBudgetAction_disappears
=== CONT  TestAccAWSBudgetsBudgetAction_basic
=== CONT  TestAccAWSBudgetsBudgetAction_disappears
--- PASS: TestAccAWSBudgetsBudgetAction_disappears (21.62s)
--- PASS: TestAccAWSBudgetsBudgetAction_basic (55.79s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	58.840s
```

##### GovCloud

```console
% make testacc TEST=./aws TESTARGS='-run=TestAccAWSBudgetsBudgetAction_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSBudgetsBudgetAction_ -timeout 180m
=== RUN   TestAccAWSBudgetsBudgetAction_basic
=== PAUSE TestAccAWSBudgetsBudgetAction_basic
=== RUN   TestAccAWSBudgetsBudgetAction_disappears
=== PAUSE TestAccAWSBudgetsBudgetAction_disappears
=== CONT  TestAccAWSBudgetsBudgetAction_basic
=== CONT  TestAccAWSBudgetsBudgetAction_disappears
=== CONT  TestAccAWSBudgetsBudgetAction_basic
    provider_test.go:734: skipping tests; partition aws-us-gov does not support budgets service
=== CONT  TestAccAWSBudgetsBudgetAction_disappears
    provider_test.go:734: skipping tests; partition aws-us-gov does not support budgets service
--- SKIP: TestAccAWSBudgetsBudgetAction_basic (1.54s)
--- SKIP: TestAccAWSBudgetsBudgetAction_disappears (1.54s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	4.596s
```

**Comment by ewbankkit** [comment]
@DrFaust92 Thanks for the contribution :tada: :clap:.
I tweaked the waiters statuses a bit to get the acceptance tests to pass.
As this is a completely new resource I am happy to merge as is and deal with issues when/if they arise.

**Pull request merged**